### PR TITLE
Update SPK recipe for CN-celeb

### PR DESCRIPTION
**Pull request by holvan**
## What did you change?

Update cnceleb/spk1 recipe based on espnet2 for speaker identification:
- Add ResNet models
- Add an exponential decay warmup scheduler
- Add an `enroll_mode` option in `spk.sh` to support enrollment with multiple utterances

---

## Why did you make this change?

To tune the models for improved performance on the CN-Celeb dataset.

---

## Is your PR small enough?

yes

### CODE REVIEW

**Comment by sw005320** [comment]
This pull request introduces new functionality for handling speaker embeddings and trials, updates inference configurations, and adds new training recipes for improved model performance. The changes focus on embedding processing, trial conversion, and training configurations for speaker recognition tasks.

### Embedding and Trial Processing Enhancements:
* Added `average_enroll_embeddings` and `main_emb_avg` functions to compute averaged embeddings for enrollment utterances and compress trial data. This improves the handling of enrollment embeddings during inference. (`egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py`) [[1]](diffhunk://#diff-a1e9c51cdab31deb915b5eb3dd5caed20ebc4f463f9caa2c33814e8b22aef281R9-R39) [[2]](diffhunk://#diff-a1e9c51cdab31deb915b5eb3dd5caed20ebc4f463f9caa2c33814e8b22aef281R85-R197)
* Updated `convert_trial.py` to support separate enrollment utterances via a new `main_sep` function, enhancing flexibility for trial data preparation. (`egs2/cnceleb/spk1/local/convert_trial.py`)

### Inference Configuration Updates:
* Introduced the `enroll_mode` parameter in `spk.sh` to support different enrollment modes, including `single`, `emb-avg`, and `score-avg`. Adjusted test set handling and embedding suffixes based on the selected mode. (`egs2/TEMPLATE/spk1/spk.sh`) [[1]](diffhunk://#diff-6c1416f444b1cc5a455109a7bb49d89bdc2e2c44123071344205b78586f03f2aR73) [[2]](diffhunk://#diff-6c1416f444b1cc5a455109a7bb49d89bdc2e2c44123071344205b78586f03f2aR199-R210) [[3]](diffhunk://#diff-6c1416f444b1cc5a455109a7bb49d89bdc2e2c44123071344205b78586f03f2aL575-R594) [[4]](diffhunk://#diff-6c1416f444b1cc5a455109a7bb49d89bdc2e2c44123071344205b78586f03f2aR614-R621)

### Training Recipes and Configuration Updates:
* Added new training configurations for ResNet34 and ResNet221 models with optimized hyperparameters, resulting in improved EER and minDCF metrics. (`egs2/cnceleb/spk1/conf/train_resnet34.yaml`, `egs2/cnceleb/spk1/conf/train_resnet221.yaml`) [[1]](diffhunk://#diff-c6eee9867197a7d69cc2a28b602029b95247ba12e6b41121d3cfb046423337fbR1-R88) [[2]](diffhunk://#diff-fc4313eb2d5b5341242ed298c1001ea80a8f4a6649bc40310f9cb5dc597f71bbR1-R89)
* Replaced the ECAPA-mel training recipe with ECAPA-TDNN, updated optimizer and scheduler settings for better performance. (`egs2/cnceleb/spk1/conf/tuning/train_ecapa_tdnn.yaml`)

### Documentation and Configuration Adjustments:
* Updated `README.md` to reflect new model results and configurations, showcasing improved performance metrics. (`egs2/cnceleb/spk1/README.md`)
* Adjusted `decode.yaml` to refine evaluation parameters, including target duration and cohort settings. (`egs2/cnceleb/spk1/conf/decode.yaml`)

These updates collectively enhance the speaker recognition pipeline, providing more robust handling of embeddings and trials, improved inference flexibility, and better training configurations for state-of-the-art models.

File: espnet2/schedulers/exponential_decay_warmup.py
Comment (sw005320) [inline espnet2/schedulers/exponential_decay_warmup.py:1]: Can you add a test to https://github.com/espnet/espnet/tree/master/test/espnet2/schedulers ?

**Comment by sw005320** [comment]
@holvan, thanks for your follow-up PR.
Can you check the copilot's review?

@Jungjee, can you review this PR?

File: egs2/cnceleb/spk1/local/data.sh [hunk 1]
```diff
@@ -123,15 +126,15 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
     # Convert CN-Celeb1
     if [ ! -d "${cnceleb1_wav_dir}" ]; then
         log "Converting CN-Celeb1 FLAC to WAV..."
-        python local/flac2wav.py --dataset_dir "${cnceleb1_flac_dir}" --out_dir "${cnceleb1_wav_dir}" --nj ${n_proc}
+        python local/flac2wav.py --dataset_dir "${cnceleb1_flac_dir}" --nj ${n_proc}
```
[the remark this answers is not in the document]
Comment (holvan) [reply egs2/cnceleb/spk1/local/data.sh:128]: The `flac2wav.py` does not have `--out_dir` argument. it automatically creates a output directory with '_wav' suffix.

**Comment by sw005320** [comment]
@Jungjee, this is a reminder.
Can you review this PR?

The other parts look good to me.

File: egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py [hunk 2]
```diff
@@ -6,6 +6,37 @@
 import torch
 
 
+def average_enroll_embeddings(embd_dir: str, enroll_list: list) -> str:
+    embd_dic = OrderedDict(np.load(embd_dir))
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py:21]: can you add a comment how embd_dir should look like?

File: egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py [hunk 3]
```diff
@@ -6,6 +6,37 @@
...
+    utts = set(embd_dic.keys())
+    enrolls = set(enroll_list)
+    tests = utts - enrolls
+    enroll_embd_dic = {}
+
+    # Gather all enroll embeddings
+    for enroll in enrolls:
+        spk = enroll.split("-")[0] + "-enroll"
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py]: are there some documentation about this rule?

File: egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py [hunk 4]
```diff
@@ -51,5 +82,117 @@ def main(args):
...
+    embd_dic = load_embeddings(embd_dir_new)
+
+    # Compress trials
+    check_list = set()
+    trial_ids_new = []
+    enrolls_new = []
+    tests_new = []
+    labels_new = []
+    for i, e in enumerate(enrolls):
+        spk = e.split("-")[0] + "-enroll"
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py]: ditto

File: egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py [hunk 5]
```diff
@@ -6,6 +6,37 @@
 import torch
```
Comment (Jungjee) [inline egs2/TEMPLATE/asr1/pyscripts/utils/spk_calculate_scores_from_embeddings.py:6]: Could you provide some comments to this script in general?

File: egs2/cnceleb/spk1/local/convert_trial.py [hunk 6]
```diff
@@ -42,14 +42,61 @@ def main(args):
             f2.write(f"{joint_key} {test_scp[test_id]}\n")
             flabel.write(f"{joint_key} {label}\n")
 
-    print(f"Trial conversion complete. Files saved in: {args.out}")
+    print(f"[Done] Trial conversion complete. Files saved in: {args.out}")
+
+
+def main_sep(args):
```
Comment (Jungjee) [inline egs2/cnceleb/spk1/local/convert_trial.py:49]: ditto, please add comments within this function in general.

**Review by Jungjee — COMMENTED** [review]
I didn't find problematic codes when going through it.
However, in general, some documentations in the form of comments, especially for data processing and enrollment processing parts, would be very helpful.

Thank you for your contribution to ESPnet-SPK!

**Comment by sw005320** [comment]
Thanks, @holvan!
